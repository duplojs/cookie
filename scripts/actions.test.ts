import { serializeCookies, parseCookies, verifyCookies } from "./actions";
import { Duplo, Response, Request } from "@duplojs/core";
import { cookiePlugin } from "./plugin";
import { Algorithm } from "./sign";

describe("actions", () => {
	const duplo = new Duplo({
		environment: "TEST",
		plugins: [cookiePlugin()],
	});

	const response = new Response(200, "ok", {})
		.setCookie("test", "test");

	const responseWithOptions = new Response(200, "ok", {})
		.setCookie("test", "test", {
			httpOnly: true,
			secure: true,
			sameSite: "strict",
		});

	it("should be implemente serializeCookies", () => {
		serializeCookies(response, {});

		expect(response).toMatchSnapshot();
	});

	it("should be implemente serializeCookies with options", () => {
		serializeCookies(responseWithOptions, {});

		expect(responseWithOptions).toMatchSnapshot();
	});

	const request = new Request({
		method: "GET",
		url: "/test",
		host: "localhost",
		origin: "http://localhost",
		headers: {
			cookie: "test=test",
		},
		matchedPath: "/test",
		params: {},
		path: "/test",
		query: {},
	});

	it("should be implemente parseCookies", () => {
		parseCookies(request, {});

		expect(request.cookies).toMatchSnapshot();
	});
});

describe("actions with secret option", () => {
	const duplo = new Duplo({
		environment: "TEST",
		plugins: [
			cookiePlugin({
				secret: "secret",
				algorithm: Algorithm.SHA256,
			}),
		],
	});

	const responseWithSecret = new Response(200, "ok", {})
		.setCookie("test", "test");

	const responseWithOptionsWithSecret = new Response(200, "ok", {})
		.setCookie("test", "test", {
			httpOnly: true,
			secure: true,
			sameSite: "strict",
		});

	it("should be implemente serializeCookies", () => {
		serializeCookies(responseWithSecret, {});

		expect(responseWithSecret).toMatchSnapshot();
	});

	it("should be implemente serializeCookies with options", () => {
		serializeCookies(responseWithOptionsWithSecret, {});

		expect(responseWithOptionsWithSecret).toMatchSnapshot();
	});

	const requestWithSignedCookies = new Request({
		method: "GET",
		url: "/test",
		host: "localhost",
		origin: "http://localhost",
		headers: {
			cookie: "test=test",
		},
		matchedPath: "/test",
		params: {},
		path: "/test",
		query: {},
	});

	it("should be implemente parseCookies", () => {
		parseCookies(requestWithSignedCookies, {
			secret: "secret",
		});

		expect(requestWithSignedCookies.cookies).toMatchSnapshot();
	});

	it("should be implemente verifyCookies", () => {
		const cookies = {
			test: "test",
			test2: undefined,
		};

		verifyCookies(cookies, "secret", Algorithm.SHA256);
	});
});
