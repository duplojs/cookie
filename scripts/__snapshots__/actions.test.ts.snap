// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`actions > should be implemente parseCookies 1`] = `
{
  "test": "test",
}
`;

exports[`actions > should be implemente serializeCookies 1`] = `
Response {
  "body": {},
  "code": 200,
  "headers": {
    "Set-Cookie": [
      "test=test; HttpOnly; Secure; SameSite=Strict",
    ],
  },
  "information": "ok",
  "keepAlive": false,
  Symbol(unique): Symbol(unique),
}
`;

exports[`actions > should be implemente serializeCookies with options 1`] = `
Response {
  "body": {},
  "code": 200,
  "headers": {
    "Set-Cookie": [
      "test=test; HttpOnly; Secure; SameSite=Strict",
    ],
  },
  "information": "ok",
  "keepAlive": false,
  Symbol(unique): Symbol(unique),
}
`;

exports[`actions with secret option > should be implemente parseCookies 1`] = `
{
  "test": "test",
}
`;

exports[`actions with secret option > should be implemente serializeCookies 1`] = `
Response {
  "body": {},
  "code": 200,
  "headers": {
    "Set-Cookie": [
      "test=test; HttpOnly; Secure; SameSite=Strict",
    ],
  },
  "information": "ok",
  "keepAlive": false,
  Symbol(unique): Symbol(unique),
}
`;

exports[`actions with secret option > should be implemente serializeCookies with options 1`] = `
Response {
  "body": {},
  "code": 200,
  "headers": {
    "Set-Cookie": [
      "test=test; HttpOnly; Secure; SameSite=Strict",
    ],
  },
  "information": "ok",
  "keepAlive": false,
  Symbol(unique): Symbol(unique),
}
`;
